### DISCUSSION TITLE
Design a better interface to allow Mixed type flows

### DISCUSSION THREAD

**Discussion by ogabrielluiz** [langflow-ai/langflow#277] [Ideas]
The objective of a flow is to build one graph that in the end turns into a Chain or AgentExecutor. 

It would be really interesting to have a way to have multiple steps.
For example:

- LLMChain that summarizes something
- Function that takes the summary and adds it to a file
- Agent that takes the output from the function and does something with it.

This would require the graphs to be directional but the nodes that make the chain should still be composition nodes.

A solution would be to create a different type of node which is a Connector. These Connectors would pass information between Composed nodes. 

Essentially, composed nodes could be seen as one entity and nodes that have a certain output could be connected to Connectors.

A few requirements I can think of right now for this work:

- Chains and Agents need a type of output that connects to a Connector node
- Since only Chains and Agents(and Connectors) can receive input, there should be an input handle on them
- Connector will have one input and one output (let's start with (str) -> str)
- The backend now must have 1 or more steps, meaning the Graph will build composed nodes and connect the composed nodes when a Connector appears. Maybe the Graph itself must be a connector between the Chat and the flow.

The flow then becomes sort of a stack starting at the Chat, going to an output which will be the input of a flow, which will output to a connector and that will output to another flow, etc.

**Comment by arunnsit** [comment]
Hi Gabriel, 

This is a great idea. Is there a simple way to extend this design to support two way conversations, for example: two models running a debate?

Talking about holistic things, how do you feel about the growth on langflow and specifically about how much versatility of langchain can be captured by langflow? Are we hitting the limits already that we need dedicated component like Connectors? or if Connectors is not something that dedicated and may be we should implement it as part of langchain and keep the simplicity of only rendering flows in langflow without heavy dedicated logic?

Looking forward to your thoughts!

Thanks

**Reply by ogabrielluiz** [reply]
Hey @arunnsit,

So, I thought some more about this.. the Connector and even more so the Graph object in LangFlow is just a bridge between LangChain and React Flow. Essentially, the Connector would be a new way to process this Graph.

To me, LangChain requires more than one Agent using Tools or one Chain to work well so if we had a way to pass information from one Chain to the other without requiring an Agent to pick the Tools, etc, we'd be expanding the functionality by a lot.

I will start experimenting with this today by creating a GraphRunner of sorts (DirectionalGraph?). A chain is a Graph it ends at the Chain object, then a new Graph starts which could be another Chain or a Connector. To do this I'd have to create a way to add an input to all objects that are type Chain (AgentExecutors inherit from Chain) so that I can connect a Chain to it and have that input be processed separately from the other params.

Adding this to LangFlow will put it a lot closer to being able to create Custom Components since all they'd need is a well-defined structure and we should be able to create it using the right nodes. With custom components there will possibly be endless ways to build the graph.

I'm sorry if I rambled hahha I'm just a bit too excited to start working on this.

**Comment by arunnsit** [comment]
Hi Gabriel,

Looking forward to this feature,  is this a priority task being pursued actively?

Thanks,
Arun

**Comment by carlosrcoelho** [comment]
Hey!

Thanks for sharing your question here. I'm sorry for not getting back to you sooner.
Langflow has gone through some major changes since then. I recommend trying out the new version for the best experience. If you have any issues, please open a new one; we'll be here to help.
